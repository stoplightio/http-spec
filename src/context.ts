import { hasRef, isLocalRef } from '@stoplight/json';
// @ts-ignore
import * as fnv from 'fnv-plus';

import type {
  AvailableContext,
  Fragment,
  IdGenerator,
  RefResolver,
  TransformerContext,
  TranslateFunction,
} from './types';

export const DEFAULT_ID_GENERATOR: IdGenerator = input => fnv.fast1a32hex(input);

export function createContext<T extends Record<string, unknown>>(
  document: T,
  resolveRef: RefResolver<T>,
  generateId: IdGenerator,
): TransformerContext<T> {
  let context: AvailableContext = 'service';
  return {
    document,
    get context() {
      return context;
    },
    set context(value) {
      context = value;
      if (value !== 'operation') {
        this.parentId = this.ids[value];
      }
    },
    maybeResolveLocalRef(target: unknown) {
      if (hasRef(target) && isLocalRef(target.$ref)) {
        try {
          return resolveRef.call(this, target);
        } catch {
          return target;
        }
      }

      return target;
    },
    generateId(value) {
      this.parentId = generateId(value);
      return this.parentId;
    },
    ids: {
      service: '',
      path: '',
      operation: '',
    },
    parentId: '',
  };
}

export function withContext<F extends TranslateFunction<any, any[]> = TranslateFunction<Fragment, unknown[]>>(
  fn: F,
): F {
  return <F>function (...args: Parameters<F>) {
    const { context, parentId } = this;
    const result = fn.apply(this, args);
    this.context = context;
    this.parentId = parentId;
    return result;
  };
}
