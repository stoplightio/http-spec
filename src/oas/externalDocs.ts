import { IExternalDocs } from '@stoplight/types';

/**
 * Extract an OpanAPI `externalDocs` object
 * @return an object that includes an `externalDocs` object with all well-formed
 * properties from the supplied object, or an empty object
 */
export function toExternalDocs(externalDocs: any): { externalDocs?: IExternalDocs } {
  if (!externalDocs || typeof externalDocs !== 'object') {
    return {};
  }

  const url = getUrl(externalDocs);
  if (url === undefined) {
    return {};
  }

  const description = getDescription(externalDocs);
  const descriptionProp = description ? { description } : {};

  return {
    externalDocs: {
      url,
      ...descriptionProp,
    },
  };
}

const ALL_WHITESPACE: RegExp = /^\s*$/;

/**
 * @returns the `url` property, or undefined if the object lacks it
 */
function getUrl(externalDocs: { url?: any } | undefined): string | undefined {
  const url = externalDocs?.url;
  if (!url || typeof url !== 'string' || ALL_WHITESPACE.test(url)) {
    return undefined;
  }

  return url;
}

/**
 * @returns the `description` property, or undefined if the object lacks it
 */
function getDescription(externalDocs: { description?: any }): string | undefined {
  const description = externalDocs.description;
  if (!description || typeof description !== 'string' || ALL_WHITESPACE.test(description)) {
    return undefined;
  }

  return description;
}
