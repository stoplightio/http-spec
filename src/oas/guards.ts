import { DeepPartial, Dictionary } from '@stoplight/types';
import { InfoObject } from 'openapi3-ts/src/model/OpenApi';
import { Info } from 'swagger-schema-official';

import { isDictionary } from '../utils';

export function hasXLogo(
  info: DeepPartial<Info | InfoObject>,
): info is DeepPartial<Info | InfoObject> & { 'x-logo': Dictionary<unknown> } {
  return isDictionary(info['x-logo']);
}
