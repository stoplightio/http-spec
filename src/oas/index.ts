export * from './operation';
export * from './types';
