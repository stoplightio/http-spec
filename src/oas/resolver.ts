import type { JsonPath } from '@stoplight/types';

import type { AvailableContext, RefResolver } from '../types';
import { resolveInlineRef } from './resolveInlineRef';

function inferContext(path: JsonPath): AvailableContext {
  if (path.length < 2 || path[0] !== 'paths') return 'service';
  if (path.length === 2 || path[3] === 'parameters' || path[3] === 'servers') return 'path';
  return 'operation';
}

const SHARED_COMPONENTS_KEYS = new WeakMap();

export function getSharedKey(value: object) {
  return SHARED_COMPONENTS_KEYS.get(value);
}

export const resolveRef: RefResolver = function (target) {
  const location: string[] = [];
  const resolved = resolveInlineRef(this.document, target.$ref, location);

  const context = inferContext(location);
  if (context !== null && this.context !== context) {
    this.context = context;
  }

  if (typeof resolved === 'object' && resolved !== null && context === 'service') {
    SHARED_COMPONENTS_KEYS.set(resolved, location[0] === 'components' ? location[2] : location[1]);
  }

  return resolved;
};
