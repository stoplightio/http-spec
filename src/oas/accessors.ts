import type { Extensions } from '@stoplight/types';

import { Fragment, TransformerContext } from '../types';
import { entries } from '../utils';
import { isValidOas2Param, isValidOas3Param, Oas2ParamBase, Oas3ParamBase, ParamBase } from './guards';
import { OasVersion } from './types';

const ROOT_EXTENSIONS = ['x-internal'];

const getIdForParameter = (param: ParamBase) => `${param.name}-${param.in}`;

export function iterateOasParams(
  this: TransformerContext,
  spec: OasVersion.OAS2,
  path: Fragment,
  operation: Fragment,
): Iterable<Oas2ParamBase>;
export function iterateOasParams(
  this: TransformerContext,
  spec: OasVersion.OAS3,
  path: Fragment,
  operation: Fragment,
): Iterable<Oas3ParamBase>;
export function* iterateOasParams(
  this: TransformerContext,
  spec: OasVersion,
  path: Fragment,
  operation: Fragment,
): Iterable<Oas2ParamBase | Oas3ParamBase> {
  const seenParams = new Set();
  const { parentId, context } = this;
  const opParams = Array.isArray(operation.parameters) ? operation.parameters : [];
  const params = [...opParams, ...(Array.isArray(path.parameters) ? path.parameters : [])];

  for (let i = 0; i < params.length; i++) {
    const param = this.maybeResolveLocalRef(params[i]);
    if (!(spec === OasVersion.OAS2 ? isValidOas2Param : isValidOas3Param)(param)) continue;

    const key = getIdForParameter(param);

    if (seenParams.has(key)) continue;
    seenParams.add(key);

    if (this.context !== 'service') {
      this.context = i < opParams.length ? 'operation' : 'path';
    }

    yield param;
  }

  this.context = context;
  this.parentId = parentId;
}

export function getExtensions(target: unknown): Extensions {
  return Object.fromEntries(entries(target).filter(([key]) => key.startsWith('x-') && !ROOT_EXTENSIONS.includes(key)));
}
