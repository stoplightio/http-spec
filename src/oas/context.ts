// @ts-expect-error: no types
import * as fnv from 'fnv-plus';

import { createContext as _createContext } from '../context';
import type { Fragment, IdGenerator } from '../types';
import { resolveRef } from './resolver';

const HASH_32_KINDS = ['http_operation', 'schema'];

function getKind(input: string) {
  const i = input.indexOf('-');
  return input.slice(0, i);
}

const DEFAULT_ID_GENERATOR: IdGenerator = input => {
  const kind = getKind(input);
  return HASH_32_KINDS.some(k => kind === k) ? fnv.fast1a32hex(input) : fnv.fast1a52hex(input);
};

export function createContext<T extends Fragment>(document: T) {
  return _createContext<T>(document, resolveRef, DEFAULT_ID_GENERATOR);
}
