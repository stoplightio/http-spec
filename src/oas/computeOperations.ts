import { IHttpOperation } from '@stoplight/types';
import { flatten, get, keys, map } from 'lodash';
import { OpenAPIObject } from 'openapi3-ts';
import { Spec } from 'swagger-schema-official';

import { HttpOperationTransformer } from '../oas/types';

const methods = ['get', 'post', 'put', 'delete', 'options', 'head', 'patch', 'trace'];

export function computeOasOperations(
  document: Spec | OpenAPIObject,
  transformer: HttpOperationTransformer<any>,
): IHttpOperation[] {
  const paths = keys(get(document, 'paths'));

  return flatten(
    map(paths, path =>
      keys(get(document, ['paths', path]))
        .filter(pathKey => methods.includes(pathKey))
        .map(method =>
          transformer({
            document,
            path,
            method,
          }),
        ),
    ),
  );
}
