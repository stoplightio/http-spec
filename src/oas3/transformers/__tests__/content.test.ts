import type { SchemaObject } from 'openapi3-ts';

import { createContext, DEFAULT_ID_GENERATOR } from '../../../context';
import { resolveRef } from '../../../oas/resolver';
import {
  translateHeaderObject as _translateHeaderObject,
  translateMediaTypeObject as _translateMediaTypeObject,
} from '../content';

const translateMediaTypeObject = (document: any, object: unknown, key: string) =>
  _translateMediaTypeObject.call(createContext(document, resolveRef, DEFAULT_ID_GENERATOR), [key, object], 0, []);

const translateHeaderObject = (object: unknown, key: string) =>
  _translateHeaderObject.call(createContext({}, resolveRef, DEFAULT_ID_GENERATOR), [key, object], 0, []);

describe('translateMediaTypeObject', () => {
  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('should gracefully handle invalid data', () => {
    expect(translateMediaTypeObject({}, null, 'mediaType')).toBeUndefined();
  });

  it('given empty object, should return nothing', () => {
    expect(translateMediaTypeObject({}, {}, 'mediaType')).toStrictEqual({
      id: expect.any(String),
      encodings: [],
      examples: [],
      mediaType: 'mediaType',
    });
  });

  it('given non-object schema, should return nothing', () => {
    expect(
      translateMediaTypeObject(
        {},
        {
          schema: 'foo',
        },
        'mediaType',
      ),
    ).toStrictEqual({
      id: expect.any(String),
      encodings: [],
      examples: [],
      mediaType: 'mediaType',
    });
  });

  it('given single example should translate to IHttpContent', () => {
    expect(
      translateMediaTypeObject(
        {},
        {
          schema: {},
          example: 'hey',
          encoding: {},
        },
        'mediaType',
      ),
    ).toMatchSnapshot({
      id: expect.any(String),
      examples: [
        {
          id: expect.any(String),
        },
      ],
      schema: {
        'x-stoplight-id': expect.any(String),
      },
    });
  });

  it('given multiple examples should translate to IHttpContent', () => {
    expect(
      translateMediaTypeObject(
        {},
        {
          schema: {},
          examples: { example: { summary: 'multi example', value: 'hey' } },
          encoding: {},
        },
        'mediaType',
      ),
    ).toMatchSnapshot({
      id: expect.any(String),
      examples: [
        {
          id: expect.any(String),
        },
      ],
      schema: {
        'x-stoplight-id': expect.any(String),
      },
    });
  });

  it('given encodings should translate each encoding to array item', () => {
    expect(
      translateMediaTypeObject(
        {},
        {
          schema: {},
          examples: {
            example: {
              id: 12133433,
              summary: 'multi example',
              value: 'hey',
            },
          },
          encoding: {
            enc1: {
              contentType: 'text/plain',
              headers: {
                h1: {},
              },
              style: 'form',
              explode: true,
              allowReserved: true,
            },
            enc2: {
              contentType: 'text/plain',
              headers: {
                h1: {},
              },
              style: 'form',
              explode: true,
              allowReserved: true,
            },
          },
        },
        'mediaType',
      ),
    ).toMatchSnapshot({
      id: expect.any(String),
      encodings: [
        {
          headers: [
            {
              id: expect.any(String),
            },
          ],
        },
        {
          headers: [
            {
              id: expect.any(String),
            },
          ],
        },
      ],
      examples: [
        {
          id: expect.any(String),
        },
      ],
      schema: {
        'x-stoplight-id': expect.any(String),
      },
    });
  });

  it('given complex nested media type object should translate correctly', () => {
    expect(
      translateMediaTypeObject(
        {},
        {
          schema: {},
          examples: { example: { summary: 'multi example', value: 'hey' } },
          encoding: {
            enc1: {
              contentType: 'text/plain',
              style: 'form',
              headers: {
                h1: {
                  description: 'descr',
                  style: 'matrix',
                  examples: {
                    a: {
                      summary: 'example summary',
                      value: 'hey',
                    },
                  },
                  content: {
                    'nested/media': {},
                  },
                },
              },
            },
          },
        },
        'mediaType',
      ),
    ).toMatchSnapshot({
      id: expect.any(String),
      encodings: [
        {
          headers: [
            {
              id: expect.any(String),
              examples: [
                {
                  id: expect.any(String),
                },
              ],
            },
          ],
        },
      ],
      examples: [
        {
          id: expect.any(String),
        },
      ],
      schema: {
        'x-stoplight-id': expect.any(String),
      },
    });
  });

  it('should skip nullish headers', () => {
    expect(
      translateMediaTypeObject(
        {},
        {
          schema: {},
          examples: { example: { summary: 'multi example', value: 'hey' } },
          encoding: {
            enc1: {
              contentType: 'text/plain',
              style: 'form',
              headers: {
                '0': null,
              },
            },
          },
        },
        'mediaType',
      ),
    ).toMatchSnapshot({
      id: expect.any(String),
      examples: [
        {
          id: expect.any(String),
        },
      ],
      schema: {
        'x-stoplight-id': expect.any(String),
      },
    });
  });

  it('given encoding with no style it should not throw an error', () => {
    const testedFunction = () => {
      translateMediaTypeObject(
        {},
        {
          schema: {},
          examples: { example: { summary: 'multi example' } },
          encoding: {
            enc1: {
              contentType: 'text/plain',
            },
          },
        },
        'mediaType',
      );
    };
    expect(() => testedFunction()).not.toThrow();
  });

  describe('schema integrity after transform', () => {
    const schema = {
      type: 'string',
      nullable: true,
      description: 'A simple string',
      example: 'hello',
      deprecated: true,
      writeOnly: true,
      readOnly: true,
      externalDocs: { url: 'http://example.com/docs', description: 'Shiny docs' },
      xml: {},
    };

    const originalSchema = JSON.parse(JSON.stringify(schema));
    const translatedObject = translateMediaTypeObject(
      {},
      {
        schema,
        example: 'hey',
        encoding: {},
      },
      'mediaType',
    )!;

    it('will not modify the original schema so it can be reused in references ', () => {
      expect(schema).toStrictEqual(originalSchema);
    });

    it('will keep the deprecated property', () => {
      expect(translatedObject.schema).toHaveProperty('deprecated', true);
    });

    it('will keep the writeOnly property', () => {
      expect(translatedObject.schema).toHaveProperty('writeOnly', true);
    });

    it('will keep the readOnly property', () => {
      expect(translatedObject.schema).toHaveProperty('readOnly', true);
    });

    it('will keep the xml property', () => {
      expect(translatedObject.schema).toHaveProperty('xml', {});
    });

    it('will keep the externalDocs property', () => {
      expect(translatedObject.schema).toHaveProperty('externalDocs', {
        url: 'http://example.com/docs',
        description: 'Shiny docs',
      });
    });
  });

  describe('schema examples', () => {
    const defaultExample = {
      name: 'default value',
    };

    describe('given response with schema with examples', () => {
      it('should translate to IHttpContent', () => {
        expect(
          translateMediaTypeObject(
            {},
            {
              schema: {
                example: defaultExample,
              },
              encoding: {},
            },
            'mediaType',
          ),
        ).toHaveProperty('examples', [{ id: expect.any(String), key: 'default', value: defaultExample }]);
      });
    });

    describe('given response with examples in media and schema objects', () => {
      it('root example should take precedence over schema example', () => {
        expect(
          translateMediaTypeObject(
            {},
            {
              schema: {
                example: defaultExample,
              },
              examples: { example: { value: { name: 'root example value' } } },
              example: {
                name: 'root default value',
              },
              encoding: {},
            },
            'mediaType',
          ),
        ).toHaveProperty('examples', [
          { id: expect.any(String), key: 'default', value: { name: 'root default value' } },
          { id: expect.any(String), key: 'example', value: { name: 'root example value' } },
        ]);
      });
    });
  });
});

describe('schema invalid', () => {
  it('type as array does not throw error', () => {
    const schema = {
      type: ['string', 'object'],
      description: 'A simple string',
      example: 'hello',
    } as unknown as SchemaObject;

    expect(() =>
      translateMediaTypeObject(
        {},
        {
          schema,
        },
        'mediaType',
      ),
    ).not.toThrow();
  });

  it('nullish values in arrays', () => {
    /*
      # Equivalent in YAML
      type: array
      items:
        -
        - object
     */
    const schema = {
      type: 'array',
      items: [null, 'object'],
    };

    expect(() =>
      translateMediaTypeObject(
        {},
        {
          schema,
        },
        'mediaType',
      ),
    ).not.toThrow();
  });

  it('nullish mapping value', () => {
    /*
     # Equivalent in YAML
     type: array
     items:
    */
    const schema = {
      type: 'array',
      items: null,
    } as any;

    expect(() =>
      translateMediaTypeObject(
        {},
        {
          schema,
        },
        'mediaType',
      ),
    ).not.toThrow();
  });
});

describe('translateHeaderObject', () => {
  it('should translate to IHttpHeaderParam', () => {
    expect(
      translateHeaderObject(
        {
          description: 'descr',
          required: true,
          deprecated: true,
          allowEmptyValue: true,
          style: 'matrix',
          explode: true,
          allowReserved: true,
          schema: {},
          examples: {
            a: {
              summary: 'example summary',
              value: 'hey',
            },
          },
          example: {},
          content: {},
        },
        'header-name',
      ),
    ).toMatchSnapshot({
      id: expect.any(String),
      examples: [
        {
          id: expect.any(String),
        },
        {
          id: expect.any(String),
        },
      ],
      schema: {
        'x-stoplight-id': expect.any(String),
      },
    });
  });

  it('should handle nullish value gracefully', () => {
    expect(translateHeaderObject(null, 'header')).toBeUndefined();
  });
});
