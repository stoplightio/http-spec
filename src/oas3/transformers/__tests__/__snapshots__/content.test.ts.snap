// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`translateMediaTypeObject given complex nested media type object should translate correctly 1`] = `
Object {
  "encodings": Array [
    Object {
      "headers": Array [
        Object {
          "content": Object {
            "nested/media": Object {},
          },
          "description": "descr",
          "encodings": Array [],
          "examples": Array [
            Object {
              "id": Any<String>,
              "key": "a",
              "summary": "example summary",
              "value": "hey",
            },
          ],
          "id": Any<String>,
          "name": "h1",
          "style": "simple",
        },
      ],
      "mediaType": "text/plain",
      "property": "enc1",
      "style": "form",
    },
  ],
  "examples": Array [
    Object {
      "id": Any<String>,
      "key": "example",
      "summary": "multi example",
      "value": "hey",
    },
  ],
  "id": Any<String>,
  "mediaType": "mediaType",
  "schema": Object {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "x-stoplight": Object {
      "id": Any<String>,
    },
  },
}
`;

exports[`translateMediaTypeObject given encodings should translate each encoding to array item 1`] = `
Object {
  "encodings": Array [
    Object {
      "allowReserved": true,
      "explode": true,
      "headers": Array [
        Object {
          "encodings": Array [],
          "examples": Array [],
          "id": Any<String>,
          "name": "h1",
          "style": "simple",
        },
      ],
      "mediaType": "text/plain",
      "property": "enc1",
      "style": "form",
    },
    Object {
      "allowReserved": true,
      "explode": true,
      "headers": Array [
        Object {
          "encodings": Array [],
          "examples": Array [],
          "id": Any<String>,
          "name": "h1",
          "style": "simple",
        },
      ],
      "mediaType": "text/plain",
      "property": "enc2",
      "style": "form",
    },
  ],
  "examples": Array [
    Object {
      "id": Any<String>,
      "key": "example",
      "summary": "multi example",
      "value": "hey",
    },
  ],
  "id": Any<String>,
  "mediaType": "mediaType",
  "schema": Object {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "x-stoplight": Object {
      "id": Any<String>,
    },
  },
}
`;

exports[`translateMediaTypeObject given multiple examples should translate to IHttpContent 1`] = `
Object {
  "encodings": Array [],
  "examples": Array [
    Object {
      "id": Any<String>,
      "key": "example",
      "summary": "multi example",
      "value": "hey",
    },
  ],
  "id": Any<String>,
  "mediaType": "mediaType",
  "schema": Object {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "x-stoplight": Object {
      "id": Any<String>,
    },
  },
}
`;

exports[`translateMediaTypeObject given single example should translate to IHttpContent 1`] = `
Object {
  "encodings": Array [],
  "examples": Array [
    Object {
      "id": Any<String>,
      "key": "default",
      "value": "hey",
    },
  ],
  "id": Any<String>,
  "mediaType": "mediaType",
  "schema": Object {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "x-stoplight": Object {
      "id": Any<String>,
    },
  },
}
`;

exports[`translateMediaTypeObject should skip nullish headers 1`] = `
Object {
  "encodings": Array [
    Object {
      "headers": Array [],
      "mediaType": "text/plain",
      "property": "enc1",
      "style": "form",
    },
  ],
  "examples": Array [
    Object {
      "id": Any<String>,
      "key": "example",
      "summary": "multi example",
      "value": "hey",
    },
  ],
  "id": Any<String>,
  "mediaType": "mediaType",
  "schema": Object {
    "$schema": "http://json-schema.org/draft-07/schema#",
    "x-stoplight": Object {
      "id": Any<String>,
    },
  },
}
`;
